<!doctype html>
<html lang="en">
<head>
<meta charset="utf-8">
<meta name="viewport" content="width=device-width, initial-scale=1, minimum-scale=1" />
<meta name="generator" content="pdoc 0.8.1" />
<title>guidemaker.cli API documentation</title>
<meta name="description" content="GuideMaker: The command line interface â€¦" />
<link href='https://cdnjs.cloudflare.com/ajax/libs/normalize/8.0.0/normalize.min.css' rel='stylesheet'>
<link href='https://cdnjs.cloudflare.com/ajax/libs/10up-sanitize.css/8.0.0/sanitize.min.css' rel='stylesheet'>
<link href="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/9.12.0/styles/github.min.css" rel="stylesheet">
<style>.flex{display:flex !important}body{line-height:1.5em}#content{padding:20px}#sidebar{padding:30px;overflow:hidden}#sidebar > *:last-child{margin-bottom:2cm}.http-server-breadcrumbs{font-size:130%;margin:0 0 15px 0}#footer{font-size:.75em;padding:5px 30px;border-top:1px solid #ddd;text-align:right}#footer p{margin:0 0 0 1em;display:inline-block}#footer p:last-child{margin-right:30px}h1,h2,h3,h4,h5{font-weight:300}h1{font-size:2.5em;line-height:1.1em}h2{font-size:1.75em;margin:1em 0 .50em 0}h3{font-size:1.4em;margin:25px 0 10px 0}h4{margin:0;font-size:105%}a{color:#058;text-decoration:none;transition:color .3s ease-in-out}a:hover{color:#e82}.title code{font-weight:bold}h2[id^="header-"]{margin-top:2em}.ident{color:#900}pre code{background:#f8f8f8;font-size:.8em;line-height:1.4em}code{background:#f2f2f1;padding:1px 4px;overflow-wrap:break-word}h1 code{background:transparent}pre{background:#f8f8f8;border:0;border-top:1px solid #ccc;border-bottom:1px solid #ccc;margin:1em 0;padding:1ex}#http-server-module-list{display:flex;flex-flow:column}#http-server-module-list div{display:flex}#http-server-module-list dt{min-width:10%}#http-server-module-list p{margin-top:0}.toc ul,#index{list-style-type:none;margin:0;padding:0}#index code{background:transparent}#index h3{border-bottom:1px solid #ddd}#index ul{padding:0}#index h4{margin-top:.6em;font-weight:bold}@media (min-width:200ex){#index .two-column{column-count:2}}@media (min-width:300ex){#index .two-column{column-count:3}}dl{margin-bottom:2em}dl dl:last-child{margin-bottom:4em}dd{margin:0 0 1em 3em}#header-classes + dl > dd{margin-bottom:3em}dd dd{margin-left:2em}dd p{margin:10px 0}.name{background:#eee;font-weight:bold;font-size:.85em;padding:5px 10px;display:inline-block;min-width:40%}.name:hover{background:#e0e0e0}.name > span:first-child{white-space:nowrap}.name.class > span:nth-child(2){margin-left:.4em}.inherited{color:#999;border-left:5px solid #eee;padding-left:1em}.inheritance em{font-style:normal;font-weight:bold}.desc h2{font-weight:400;font-size:1.25em}.desc h3{font-size:1em}.desc dt code{background:inherit}.source summary,.git-link-div{color:#666;text-align:right;font-weight:400;font-size:.8em;text-transform:uppercase}.source summary > *{white-space:nowrap;cursor:pointer}.git-link{color:inherit;margin-left:1em}.source pre{max-height:500px;overflow:auto;margin:0}.source pre code{font-size:12px;overflow:visible}.hlist{list-style:none}.hlist li{display:inline}.hlist li:after{content:',\2002'}.hlist li:last-child:after{content:none}.hlist .hlist{display:inline;padding-left:1em}img{max-width:100%}.admonition{padding:.1em .5em;margin-bottom:1em}.admonition-title{font-weight:bold}.admonition.note,.admonition.info,.admonition.important{background:#aef}.admonition.todo,.admonition.versionadded,.admonition.tip,.admonition.hint{background:#dfd}.admonition.warning,.admonition.versionchanged,.admonition.deprecated{background:#fd4}.admonition.error,.admonition.danger,.admonition.caution{background:lightpink}</style>
<style media="screen and (min-width: 700px)">@media screen and (min-width:700px){#sidebar{width:30%;height:100vh;overflow:auto;position:sticky;top:0}#content{width:70%;max-width:100ch;padding:3em 4em;border-left:1px solid #ddd}pre code{font-size:1em}.item .name{font-size:1em}main{display:flex;flex-direction:row-reverse;justify-content:flex-end}.toc ul ul,#index ul{padding-left:1.5em}.toc > ul > li{margin-top:.5em}}</style>
<style media="print">@media print{#sidebar h1{page-break-before:always}.source{display:none}}@media print{*{background:transparent !important;color:#000 !important;box-shadow:none !important;text-shadow:none !important}a[href]:after{content:" (" attr(href) ")";font-size:90%}a[href][title]:after{content:none}abbr[title]:after{content:" (" attr(title) ")"}.ir a:after,a[href^="javascript:"]:after,a[href^="#"]:after{content:""}pre,blockquote{border:1px solid #999;page-break-inside:avoid}thead{display:table-header-group}tr,img{page-break-inside:avoid}img{max-width:100% !important}@page{margin:0.5cm}p,h2,h3{orphans:3;widows:3}h1,h2,h3,h4,h5,h6{page-break-after:avoid}}</style>
</head>
<body>
<main>
<article id="content">
<header>
<h1 class="title">Module <code>guidemaker.cli</code></h1>
</header>
<section id="section-intro">
<p>GuideMaker: The command line interface</p>
<p>A command line tool to globally design guide RNAs for any CRISPR-Cas system in any genome</p>
<details class="source">
<summary>
<span>Expand source code</span>
</summary>
<pre><code class="python">&#34;&#34;&#34;GuideMaker: The command line interface

A command line tool to globally design guide RNAs for any CRISPR-Cas system in any genome
&#34;&#34;&#34;
import logging
import argparse
import tempfile
import shutil
import os

import pybedtools
from Bio import SeqIO

import guidemaker

def myparser():
    parser = argparse.ArgumentParser(description=&#39;GuideMaker: globally design guide RNAs for any CRISPR-Cas system in any genome&#39;)
    parser.add_argument(&#39;--genbank&#39;, &#39;-i&#39;, nargs=&#39;+&#39;, type=str, required=True, help=&#39;One or more genbank .gbk  or gzipped .gbk files for a single genome&#39;)
    parser.add_argument(&#39;--pamseq&#39;, &#39;-p&#39;, type=str, required=True, help=&#39;A short PAM motif to search for, it may use IUPAC ambiguous alphabet&#39;)
    parser.add_argument(&#39;--outfile&#39;, &#39;-o&#39;, type=str, required=True, help=&#39;The table of PAM sites and data&#39;)
    parser.add_argument(&#39;--pam_orientation&#39;, &#39;-r&#39;, choices=[&#39;5prime&#39;, &#39;3prime&#39;], default=&#39;5prime&#39;, help=&#34;PAM position relative to target: 5prime: [PAM][target], 3prime: [target][PAM]. For example, Cas9 is 3prime&#34;)
    parser.add_argument(&#39;--guidelength&#39;, &#39;-l&#39;, type=int, default=20, choices=range(10, 28, 1), metavar=&#34;[10-27]&#34; ,help=&#39;Length of the guide sequence&#39;)
    parser.add_argument(&#39;--strand&#39;, &#39;-s&#39;, choices=[&#39;forward&#39;,&#39;reverse&#39;, &#39;both&#39;], default=&#39;both&#39;, help=&#39;Strand of DNA to search&#39;)
    parser.add_argument(&#39;--lcp&#39;, type=int, default=10,choices=range(0, 28, 1), metavar=&#34;[0-27]&#34;, help=&#39;Length of the guide closest to  the PAM required to be unique&#39;)
    parser.add_argument(&#39;--dist&#39;, type=int, choices=range(1, 6, 1),  metavar=&#34;[1-5]&#34;, default=2, help=&#39;Minimum hamming distance from any other potential guide&#39;)
    parser.add_argument(&#39;--before&#39;, type=int, default=100, choices=range(1, 501, 1), metavar=&#34;[1-500]&#34;,
                        help=&#39;keep guides this far in front of a feature&#39;)
    parser.add_argument(&#39;--into&#39;, type=int, default=200, choices=range(1, 501, 1),metavar=&#34;[1-500]&#34;,
                        help=&#39;keep guides this far inside (past the start site)of a feature&#39;)
    parser.add_argument(&#39;--knum&#39;, type=int, default=3, choices=range(2, 20, 1), metavar=&#34;[2-20]&#34;, help=&#39;how many sequences similar to the guide to report&#39;)
    parser.add_argument(&#39;--threads&#39;, help=&#39;The number of cpu threads to use&#39;, type=int, default=2)
    parser.add_argument(&#39;--log&#39;, help=&#34;Log file&#34;, default=&#34;guidemaker.log&#34;)
    parser.add_argument(&#39;--tempdir&#39;, help=&#39;The temp file directory&#39;, default=None)
    parser.add_argument(&#39;-V&#39;, &#39;--version&#39;, action=&#39;version&#39;, version=&#34;%(prog)s (&#34; + guidemaker.__version__ + &#34;)&#34;)
    return parser

def parserval(args):
    assert(args.lcp &lt;= args.guidelength), &#34;The length of sequence near the PAM the is unique (lcp) must be less than the guide length&#34;
    assert(1 &lt; len(args.pamseq) &lt; 7 ), &#34;The length of the PAM sequence must be between 2-6&#34;

def _logger_setup(logfile):
    &#34;&#34;&#34;Set up logging to a logfile and the terminal standard out.

    Args:
        logfile (str): Log file

    &#34;&#34;&#34;
    try:
        logging.basicConfig(level=logging.DEBUG,
                            format=&#39;%(asctime)s %(name)-12s %(levelname)-8s %(message)s&#39;,
                            datefmt=&#39;%m-%d %H:%M&#39;,
                            filename=logfile,
                            filemode=&#39;w&#39;)
        # define a Handler which writes INFO messages or higher to the sys.stderr
        console = logging.StreamHandler()
        console.setLevel(logging.INFO)
        # set a format which is simpler for console use
        formatter = logging.Formatter(&#39;%(asctime)s: %(levelname)-8s %(message)s&#39;)
        # tell the handler to use this format
        console.setFormatter(formatter)
        # add the handler to the root logger
        logging.getLogger(&#39;&#39;).addHandler(console)
    except Exception as e:
        print(&#34;An error occurred setting up logging&#34;)
        raise e


def main(args=None):
    &#34;&#34;&#34;Run The complete GuideMaker workflow.

    &#34;&#34;&#34;
    # Set up logging
    parser = myparser()
    if not args:
        args = parser.parse_args()
    parserval(args)

    _logger_setup(args.log)
    try:

        if args.tempdir:
            if not os.path.exists(args.tempdir):
                logging.warning(&#34;Specified location for tempfile (%s) does not \
                                 exist, using default location.&#34; % tempdir )
                tempdir = tempfile.mkdtemp(prefix=&#39;guidemaker_&#39;)
        else:
            tempdir = tempfile.mkdtemp(prefix=&#39;guidemaker_&#39;, dir=args.tempdir)
            pybedtools.helpers.set_tempdir(tempdir)
        logging.info(&#34;Temp directory is: %s&#34; % (tempdir))
        logging.info(&#34;Writing fasta file from genbank file(s)&#34;)

        fastapath = guidemaker.get_fastas(args.genbank, tempdir=tempdir)
        logging.info(&#34;Identifying PAM sites in the genome&#34;)
        pamobj = guidemaker.core.Pam(args.pamseq, args.pam_orientation)
        seq_record_iter = SeqIO.parse(fastapath, &#34;fasta&#34;)
        pamtargets = pamobj.find_targets(seq_record_iter=seq_record_iter, strand=&#34;both&#34;, target_len=args.guidelength)
        tl = guidemaker.core.TargetList(targets=pamtargets, lcp=args.lcp, hammingdist=args.dist, knum=args.knum)
        logging.info(&#34;Identifing guides that are unique near the PAM site&#34;)
        tl.find_unique_near_pam()
        logging.info(&#34;Indexing all potential guide sites: %s. This is the longest step.&#34; % len(tl.targets))
        tl.create_index(num_threads=args.threads)
        logging.info(&#34;Identifying guides that have a hamming distance &lt;= %s to all other potential guides&#34;, str(args.dist))
        tl.get_neighbors(num_threads=args.threads)
        logging.info(&#34;Formatting data for BedTools&#34;)
        tf_df = tl.export_bed()
        logging.info(&#34;Create GuideMaker Annotation object&#34;)
        anno = guidemaker.core.Annotation(genbank_list=args.genbank,
                                           target_bed_df=tf_df)
        logging.info(&#34;Identify genomic features&#34;)
        anno._get_genbank_features()
        logging.info(&#34;Find genomic features closest the guides&#34;)
        anno._get_nearby_features()
        logging.info(&#34;Select guides that start between +%s and -%s of a feature start&#34; % (args.before, args.into))
        anno._filter_features(before_feat=args.before, after_feat=args.into)
        logging.info(&#34;Select description columns&#34;)
        anno._get_qualifiers()
        logging.info(&#34;Format the output&#34;)
        prettydf = anno._format_guide_table(tl)
        prettydf.to_csv(args.outfile)
        logging.info(&#34;guidemaker completed, results are at %s&#34; % args.outfile)
        logging.info(&#34;PAM sequence: %s&#34; % args.pamseq)
        logging.info(&#34;PAM orientation: %s&#34; % args.pam_orientation)
        logging.info(&#34;Genome strand(s) searched: %s&#34; % args.strand)
        logging.info(&#34;Total PAM sites considered: %d&#34; % len(tl.targets))
        logging.info(&#34;Guide RNA candidates found: %d&#34; % len(prettydf))

    except Exception as e:
        logging.error(&#34;GuideMaker terminated with errors. See the log file for details.&#34;)
        logging.error(e)
        raise SystemExit(1)
    finally:
        try:
            shutil.rmtree(tempdir)
        except UnboundLocalError:
            raise SystemExit(1)
        except AttributeError:
            raise SystemExit(1)</code></pre>
</details>
</section>
<section>
</section>
<section>
</section>
<section>
<h2 class="section-title" id="header-functions">Functions</h2>
<dl>
<dt id="guidemaker.cli.main"><code class="name flex">
<span>def <span class="ident">main</span></span>(<span>args=None)</span>
</code></dt>
<dd>
<div class="desc"><p>Run The complete GuideMaker workflow.</p></div>
<details class="source">
<summary>
<span>Expand source code</span>
</summary>
<pre><code class="python">def main(args=None):
    &#34;&#34;&#34;Run The complete GuideMaker workflow.

    &#34;&#34;&#34;
    # Set up logging
    parser = myparser()
    if not args:
        args = parser.parse_args()
    parserval(args)

    _logger_setup(args.log)
    try:

        if args.tempdir:
            if not os.path.exists(args.tempdir):
                logging.warning(&#34;Specified location for tempfile (%s) does not \
                                 exist, using default location.&#34; % tempdir )
                tempdir = tempfile.mkdtemp(prefix=&#39;guidemaker_&#39;)
        else:
            tempdir = tempfile.mkdtemp(prefix=&#39;guidemaker_&#39;, dir=args.tempdir)
            pybedtools.helpers.set_tempdir(tempdir)
        logging.info(&#34;Temp directory is: %s&#34; % (tempdir))
        logging.info(&#34;Writing fasta file from genbank file(s)&#34;)

        fastapath = guidemaker.get_fastas(args.genbank, tempdir=tempdir)
        logging.info(&#34;Identifying PAM sites in the genome&#34;)
        pamobj = guidemaker.core.Pam(args.pamseq, args.pam_orientation)
        seq_record_iter = SeqIO.parse(fastapath, &#34;fasta&#34;)
        pamtargets = pamobj.find_targets(seq_record_iter=seq_record_iter, strand=&#34;both&#34;, target_len=args.guidelength)
        tl = guidemaker.core.TargetList(targets=pamtargets, lcp=args.lcp, hammingdist=args.dist, knum=args.knum)
        logging.info(&#34;Identifing guides that are unique near the PAM site&#34;)
        tl.find_unique_near_pam()
        logging.info(&#34;Indexing all potential guide sites: %s. This is the longest step.&#34; % len(tl.targets))
        tl.create_index(num_threads=args.threads)
        logging.info(&#34;Identifying guides that have a hamming distance &lt;= %s to all other potential guides&#34;, str(args.dist))
        tl.get_neighbors(num_threads=args.threads)
        logging.info(&#34;Formatting data for BedTools&#34;)
        tf_df = tl.export_bed()
        logging.info(&#34;Create GuideMaker Annotation object&#34;)
        anno = guidemaker.core.Annotation(genbank_list=args.genbank,
                                           target_bed_df=tf_df)
        logging.info(&#34;Identify genomic features&#34;)
        anno._get_genbank_features()
        logging.info(&#34;Find genomic features closest the guides&#34;)
        anno._get_nearby_features()
        logging.info(&#34;Select guides that start between +%s and -%s of a feature start&#34; % (args.before, args.into))
        anno._filter_features(before_feat=args.before, after_feat=args.into)
        logging.info(&#34;Select description columns&#34;)
        anno._get_qualifiers()
        logging.info(&#34;Format the output&#34;)
        prettydf = anno._format_guide_table(tl)
        prettydf.to_csv(args.outfile)
        logging.info(&#34;guidemaker completed, results are at %s&#34; % args.outfile)
        logging.info(&#34;PAM sequence: %s&#34; % args.pamseq)
        logging.info(&#34;PAM orientation: %s&#34; % args.pam_orientation)
        logging.info(&#34;Genome strand(s) searched: %s&#34; % args.strand)
        logging.info(&#34;Total PAM sites considered: %d&#34; % len(tl.targets))
        logging.info(&#34;Guide RNA candidates found: %d&#34; % len(prettydf))

    except Exception as e:
        logging.error(&#34;GuideMaker terminated with errors. See the log file for details.&#34;)
        logging.error(e)
        raise SystemExit(1)
    finally:
        try:
            shutil.rmtree(tempdir)
        except UnboundLocalError:
            raise SystemExit(1)
        except AttributeError:
            raise SystemExit(1)</code></pre>
</details>
</dd>
<dt id="guidemaker.cli.myparser"><code class="name flex">
<span>def <span class="ident">myparser</span></span>(<span>)</span>
</code></dt>
<dd>
<div class="desc"></div>
<details class="source">
<summary>
<span>Expand source code</span>
</summary>
<pre><code class="python">def myparser():
    parser = argparse.ArgumentParser(description=&#39;GuideMaker: globally design guide RNAs for any CRISPR-Cas system in any genome&#39;)
    parser.add_argument(&#39;--genbank&#39;, &#39;-i&#39;, nargs=&#39;+&#39;, type=str, required=True, help=&#39;One or more genbank .gbk  or gzipped .gbk files for a single genome&#39;)
    parser.add_argument(&#39;--pamseq&#39;, &#39;-p&#39;, type=str, required=True, help=&#39;A short PAM motif to search for, it may use IUPAC ambiguous alphabet&#39;)
    parser.add_argument(&#39;--outfile&#39;, &#39;-o&#39;, type=str, required=True, help=&#39;The table of PAM sites and data&#39;)
    parser.add_argument(&#39;--pam_orientation&#39;, &#39;-r&#39;, choices=[&#39;5prime&#39;, &#39;3prime&#39;], default=&#39;5prime&#39;, help=&#34;PAM position relative to target: 5prime: [PAM][target], 3prime: [target][PAM]. For example, Cas9 is 3prime&#34;)
    parser.add_argument(&#39;--guidelength&#39;, &#39;-l&#39;, type=int, default=20, choices=range(10, 28, 1), metavar=&#34;[10-27]&#34; ,help=&#39;Length of the guide sequence&#39;)
    parser.add_argument(&#39;--strand&#39;, &#39;-s&#39;, choices=[&#39;forward&#39;,&#39;reverse&#39;, &#39;both&#39;], default=&#39;both&#39;, help=&#39;Strand of DNA to search&#39;)
    parser.add_argument(&#39;--lcp&#39;, type=int, default=10,choices=range(0, 28, 1), metavar=&#34;[0-27]&#34;, help=&#39;Length of the guide closest to  the PAM required to be unique&#39;)
    parser.add_argument(&#39;--dist&#39;, type=int, choices=range(1, 6, 1),  metavar=&#34;[1-5]&#34;, default=2, help=&#39;Minimum hamming distance from any other potential guide&#39;)
    parser.add_argument(&#39;--before&#39;, type=int, default=100, choices=range(1, 501, 1), metavar=&#34;[1-500]&#34;,
                        help=&#39;keep guides this far in front of a feature&#39;)
    parser.add_argument(&#39;--into&#39;, type=int, default=200, choices=range(1, 501, 1),metavar=&#34;[1-500]&#34;,
                        help=&#39;keep guides this far inside (past the start site)of a feature&#39;)
    parser.add_argument(&#39;--knum&#39;, type=int, default=3, choices=range(2, 20, 1), metavar=&#34;[2-20]&#34;, help=&#39;how many sequences similar to the guide to report&#39;)
    parser.add_argument(&#39;--threads&#39;, help=&#39;The number of cpu threads to use&#39;, type=int, default=2)
    parser.add_argument(&#39;--log&#39;, help=&#34;Log file&#34;, default=&#34;guidemaker.log&#34;)
    parser.add_argument(&#39;--tempdir&#39;, help=&#39;The temp file directory&#39;, default=None)
    parser.add_argument(&#39;-V&#39;, &#39;--version&#39;, action=&#39;version&#39;, version=&#34;%(prog)s (&#34; + guidemaker.__version__ + &#34;)&#34;)
    return parser</code></pre>
</details>
</dd>
<dt id="guidemaker.cli.parserval"><code class="name flex">
<span>def <span class="ident">parserval</span></span>(<span>args)</span>
</code></dt>
<dd>
<div class="desc"></div>
<details class="source">
<summary>
<span>Expand source code</span>
</summary>
<pre><code class="python">def parserval(args):
    assert(args.lcp &lt;= args.guidelength), &#34;The length of sequence near the PAM the is unique (lcp) must be less than the guide length&#34;
    assert(1 &lt; len(args.pamseq) &lt; 7 ), &#34;The length of the PAM sequence must be between 2-6&#34;</code></pre>
</details>
</dd>
</dl>
</section>
<section>
</section>
</article>
<nav id="sidebar">
<h1>Index</h1>
<div class="toc">
<ul></ul>
</div>
<ul id="index">
<li><h3>Super-module</h3>
<ul>
<li><code><a title="guidemaker" href="index.html">guidemaker</a></code></li>
</ul>
</li>
<li><h3><a href="#header-functions">Functions</a></h3>
<ul class="">
<li><code><a title="guidemaker.cli.main" href="#guidemaker.cli.main">main</a></code></li>
<li><code><a title="guidemaker.cli.myparser" href="#guidemaker.cli.myparser">myparser</a></code></li>
<li><code><a title="guidemaker.cli.parserval" href="#guidemaker.cli.parserval">parserval</a></code></li>
</ul>
</li>
</ul>
</nav>
</main>
<footer id="footer">
<p>Generated by <a href="https://pdoc3.github.io/pdoc"><cite>pdoc</cite> 0.8.1</a>.</p>
</footer>
<script src="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/9.12.0/highlight.min.js"></script>
<script>hljs.initHighlightingOnLoad()</script>
</body>
</html>