<!doctype html>
<html lang="en">
<head>
<meta charset="utf-8">
<meta name="viewport" content="width=device-width, initial-scale=1, minimum-scale=1" />
<meta name="generator" content="pdoc 0.10.0" />
<title>guidemaker.data.app API documentation</title>
<meta name="description" content="guidemaker/app.py A configuration template fo a streamlit web application version of Guidemaker." />
<link rel="preload stylesheet" as="style" href="https://cdnjs.cloudflare.com/ajax/libs/10up-sanitize.css/11.0.1/sanitize.min.css" integrity="sha256-PK9q560IAAa6WVRRh76LtCaI8pjTJ2z11v0miyNNjrs=" crossorigin>
<link rel="preload stylesheet" as="style" href="https://cdnjs.cloudflare.com/ajax/libs/10up-sanitize.css/11.0.1/typography.min.css" integrity="sha256-7l/o7C8jubJiy74VsKTidCy1yBkRtiUGbVkYBylBqUg=" crossorigin>
<link rel="stylesheet preload" as="style" href="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/10.1.1/styles/github.min.css" crossorigin>
<style>:root{--highlight-color:#fe9}.flex{display:flex !important}body{line-height:1.5em}#content{padding:20px}#sidebar{padding:30px;overflow:hidden}#sidebar > *:last-child{margin-bottom:2cm}.http-server-breadcrumbs{font-size:130%;margin:0 0 15px 0}#footer{font-size:.75em;padding:5px 30px;border-top:1px solid #ddd;text-align:right}#footer p{margin:0 0 0 1em;display:inline-block}#footer p:last-child{margin-right:30px}h1,h2,h3,h4,h5{font-weight:300}h1{font-size:2.5em;line-height:1.1em}h2{font-size:1.75em;margin:1em 0 .50em 0}h3{font-size:1.4em;margin:25px 0 10px 0}h4{margin:0;font-size:105%}h1:target,h2:target,h3:target,h4:target,h5:target,h6:target{background:var(--highlight-color);padding:.2em 0}a{color:#058;text-decoration:none;transition:color .3s ease-in-out}a:hover{color:#e82}.title code{font-weight:bold}h2[id^="header-"]{margin-top:2em}.ident{color:#900}pre code{background:#f8f8f8;font-size:.8em;line-height:1.4em}code{background:#f2f2f1;padding:1px 4px;overflow-wrap:break-word}h1 code{background:transparent}pre{background:#f8f8f8;border:0;border-top:1px solid #ccc;border-bottom:1px solid #ccc;margin:1em 0;padding:1ex}#http-server-module-list{display:flex;flex-flow:column}#http-server-module-list div{display:flex}#http-server-module-list dt{min-width:10%}#http-server-module-list p{margin-top:0}.toc ul,#index{list-style-type:none;margin:0;padding:0}#index code{background:transparent}#index h3{border-bottom:1px solid #ddd}#index ul{padding:0}#index h4{margin-top:.6em;font-weight:bold}@media (min-width:200ex){#index .two-column{column-count:2}}@media (min-width:300ex){#index .two-column{column-count:3}}dl{margin-bottom:2em}dl dl:last-child{margin-bottom:4em}dd{margin:0 0 1em 3em}#header-classes + dl > dd{margin-bottom:3em}dd dd{margin-left:2em}dd p{margin:10px 0}.name{background:#eee;font-weight:bold;font-size:.85em;padding:5px 10px;display:inline-block;min-width:40%}.name:hover{background:#e0e0e0}dt:target .name{background:var(--highlight-color)}.name > span:first-child{white-space:nowrap}.name.class > span:nth-child(2){margin-left:.4em}.inherited{color:#999;border-left:5px solid #eee;padding-left:1em}.inheritance em{font-style:normal;font-weight:bold}.desc h2{font-weight:400;font-size:1.25em}.desc h3{font-size:1em}.desc dt code{background:inherit}.source summary,.git-link-div{color:#666;text-align:right;font-weight:400;font-size:.8em;text-transform:uppercase}.source summary > *{white-space:nowrap;cursor:pointer}.git-link{color:inherit;margin-left:1em}.source pre{max-height:500px;overflow:auto;margin:0}.source pre code{font-size:12px;overflow:visible}.hlist{list-style:none}.hlist li{display:inline}.hlist li:after{content:',\2002'}.hlist li:last-child:after{content:none}.hlist .hlist{display:inline;padding-left:1em}img{max-width:100%}td{padding:0 .5em}.admonition{padding:.1em .5em;margin-bottom:1em}.admonition-title{font-weight:bold}.admonition.note,.admonition.info,.admonition.important{background:#aef}.admonition.todo,.admonition.versionadded,.admonition.tip,.admonition.hint{background:#dfd}.admonition.warning,.admonition.versionchanged,.admonition.deprecated{background:#fd4}.admonition.error,.admonition.danger,.admonition.caution{background:lightpink}</style>
<style media="screen and (min-width: 700px)">@media screen and (min-width:700px){#sidebar{width:30%;height:100vh;overflow:auto;position:sticky;top:0}#content{width:70%;max-width:100ch;padding:3em 4em;border-left:1px solid #ddd}pre code{font-size:1em}.item .name{font-size:1em}main{display:flex;flex-direction:row-reverse;justify-content:flex-end}.toc ul ul,#index ul{padding-left:1.5em}.toc > ul > li{margin-top:.5em}}</style>
<style media="print">@media print{#sidebar h1{page-break-before:always}.source{display:none}}@media print{*{background:transparent !important;color:#000 !important;box-shadow:none !important;text-shadow:none !important}a[href]:after{content:" (" attr(href) ")";font-size:90%}a[href][title]:after{content:none}abbr[title]:after{content:" (" attr(title) ")"}.ir a:after,a[href^="javascript:"]:after,a[href^="#"]:after{content:""}pre,blockquote{border:1px solid #999;page-break-inside:avoid}thead{display:table-header-group}tr,img{page-break-inside:avoid}img{max-width:100% !important}@page{margin:0.5cm}p,h2,h3{orphans:3;widows:3}h1,h2,h3,h4,h5,h6{page-break-after:avoid}}</style>
<script defer src="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/10.1.1/highlight.min.js" integrity="sha256-Uv3H6lx7dJmRfRvH8TH6kJD1TSK1aFcwgx+mdg3epi8=" crossorigin></script>
<script>window.addEventListener('DOMContentLoaded', () => hljs.initHighlighting())</script>
</head>
<body>
<main>
<article id="content">
<header>
<h1 class="title">Module <code>guidemaker.data.app</code></h1>
</header>
<section id="section-intro">
<p>guidemaker/app.py A configuration template fo a streamlit web application version of Guidemaker.</p>
<details class="source">
<summary>
<span>Expand source code</span>
</summary>
<pre><code class="python">&#34;&#34;&#34;guidemaker/app.py A configuration template fo a streamlit web application version of Guidemaker.&#34;&#34;&#34;

import os
import subprocess
import base64
import pathlib
from pathlib import Path
import uuid
from uuid import uuid4
from contextlib import contextmanager
import shutil

import pandas as pd
import altair as alt
from PIL import Image
import streamlit as st
from streamlit_tags import st_tags_sidebar
from copy import deepcopy

import guidemaker

apptitle = &#39;GuideMaker&#39;
st.set_page_config(page_title=apptitle, page_icon=&#34;:eyeglasses:&#34;)

st.sidebar.markdown(&#34;## Select Parameters to Design gRNAs&#34;)


DATA_DIR = os.path.dirname(os.path.abspath(__file__))  # This is your Project Root

# @contextmanager
# def genome_connect(db_bytes):
#     &#34;&#34;&#34;Write input genome to local disk and clean after using.&#34;&#34;&#34;
#     fp = Path(str(uuid4()))
#     with open(&#34;input.gbk&#34;, &#39;wb&#39;) as fp:
#         for i in db_bytes:
#             if guidemaker.is_gzip(i):
#                 with zip.open(i, &#39;rt&#39;) as f:
#                     fp = bytearray(f)
#                     file.write(fp)
#             else:
#                 with open(i, &#39;r&#39;) as f:
#                     fp = bytearray(f)
#                     file.write(fp)
#     conn = str(fp)
#     try:
#         yield conn
#     finally:
#         pass


@contextmanager
def genome_connect(db_bytes):
    &#34;&#34;&#34;Write input genome to local disk and clean after using.&#34;&#34;&#34;
    fp = Path(str(uuid4()))
    with open(&#39;input.gbk&#39;, &#39;wb&#39;) as file:
        for i in db_bytes:
            fp = bytearray(i)
            file.write(fp)
    conn = str(fp)
    try:
        yield conn
    finally:
        pass

@contextmanager
def fasta_connect(db_bytes):
    &#34;&#34;&#34;Write input genome to local disk and clean after using.&#34;&#34;&#34;
    fp = Path(str(uuid4()))
    with open(&#39;input.fasta&#39;, &#39;wb&#39;) as file:
        for i in db_bytes:
            fp = bytearray(i)
            file.write(fp)
    conn = str(fp)
    try:
        yield conn
    finally:
        pass

@contextmanager
def gff_connect(db_bytes):
    &#34;&#34;&#34;Write input genome to local disk and clean after using.&#34;&#34;&#34;
    fp = Path(str(uuid4()))
    with open(&#39;input.gff&#39;, &#39;wb&#39;) as file:
        for i in db_bytes:
            fp = bytearray(i)
            file.write(fp)
    conn = str(fp)
    try:
        yield conn
    finally:
        pass

@st.cache(suppress_st_warning=True)
def run_command(args):
    &#34;&#34;&#34;Run command, transfer stdout/stderr back into Streamlit and manage error.&#34;&#34;&#34;
    st.info(f&#34;Running:: &#39;{&#39; &#39;.join(args)}&#39;&#34;)
    result = subprocess.run(args, capture_output=True, text=True)
    try:
        result.check_returncode()
        # st.info(result.stdout)
        # st.text(result.stderr)
    except subprocess.CalledProcessError as e:
        st.error(result.stderr)
        raise e


def get_binary_file_downloader_html(bin_file, file_label=&#39;File&#39;):
    &#34;&#34;&#34;Binary file downloader in html format.&#34;&#34;&#34;
    with open(bin_file, &#39;rb&#39;) as f:
        data = f.read()
    bin_str = base64.b64encode(data).decode()
    href = f&#39;&lt;a href=&#34;data:application/octet-stream;base64,{bin_str}&#34; download=&#34;{os.path.basename(bin_file)}&#34;&gt;{file_label}&lt;/a&gt;&#39;
    return href


def read_markdown_file(markdown_file):
    &#34;&#34;&#34;Read markdown file.&#34;&#34;&#34;
    return Path(markdown_file).read_text()


def guidemakerplot(df):
    &#34;&#34;&#34;Returns guidemaker plot describing PAM targets.&#34;&#34;&#34;
    source = df
    brush = alt.selection(type=&#39;interval&#39;, encodings=[&#39;x&#39;])
    binnum = int(round(source[&#39;Feature end&#39;].max() / 200, 0))
    display_info = source.columns.tolist()

    # Feature density
    densityF = alt.Chart(source).transform_density(
    &#39;Feature start&#39;,
    as_=[&#39;Feature start&#39;, &#39;Feature Density&#39;],
    extent=[1, source[&#39;Feature end&#39;].max()],
    bandwidth=binnum,
    ).mark_area(color=&#39;black&#39;, opacity=0.6).encode(
    x=alt.X(&#39;Feature start&#39;, axis=alt.Axis(title=&#39;Genome Coordinates (bp)&#39;, tickCount=5)),
    y=&#39;Feature Density:Q&#39;,
    ).properties(height=50)

    # gRNA density
    densityg = alt.Chart(source).transform_density(
    &#39;Guide start&#39;,
    as_=[&#39;Guide start&#39;, &#39;Guide Density&#39;],
    extent=[1, source[&#39;Feature end&#39;].max()],
    bandwidth=binnum,
    ).mark_area(color=&#39;pink&#39;, opacity=0.6).encode(
    x=alt.X(&#39;Guide start&#39;, axis=alt.Axis(title=&#39;Genome Coordinates (bp)&#39;, tickCount=5)),
    y=&#39;Guide Density:Q&#39;,
    ).properties(height=50).add_selection(brush)

    # locus tag
    locus = alt.Chart(source).mark_bar(cornerRadiusTopLeft=3, cornerRadiusTopRight=3).encode(
    x=&#39;count(locus_tag):Q&#39;,
    y=alt.Y(&#39;locus_tag&#39;, axis=alt.Axis(title=&#39;Locus&#39;)),
    color=&#39;PAM:N&#39;,
    tooltip=display_info
    ).transform_filter(
    brush
    ).interactive().properties(height=500)

    gmplot = densityF &amp; densityg &amp; locus
    return gmplot




def main(arglist: list = None):
    &#34;&#34;&#34;Run web App.&#34;&#34;&#34;
    header = &#34;GuideMaker&#34;
    subheader = &#34;Software to design CRISPR-Cas guide RNA pools in non-model genomes 🦠 🧬&#34;
    st.markdown(f&#39;&lt;strong style=&#34;font-family:Hoefler Text;font-size: 36px;color: #0021A5&#34;&gt;{header}&lt;/strong&gt;&#39;,
                unsafe_allow_html=True)
    st.markdown(
        f&#39;&lt;strong style=&#34;font-family:Hoefler Text;font-size: 18px;color: #FA4616&#34;&gt;{subheader}&lt;/strong&gt;&#39;, unsafe_allow_html=True)

    st.markdown(&#34;---&#34;)

    sessionID = str(uuid.uuid1())
    # st.write(sessionID)
    logfilename = sessionID + &#34;_log.txt&#34;

    # Create a downloads directory within the streamlit static asset directory
    # and write output files to it. Binary downloader as html file  has limited
    # file size for conversion. So, we need to write it to local folder.
    STREAMLIT_STATIC_PATH = pathlib.Path(st.__path__[0]) / &#39;static&#39;
    DOWNLOADS_PATH = (STREAMLIT_STATIC_PATH / &#34;downloads&#34;)
    if not DOWNLOADS_PATH.is_dir():
        DOWNLOADS_PATH.mkdir()

    # Define input parameters and widgets

    multiple_files_gbk = st.sidebar.file_uploader(&#34;Upload one or more Genome file [ .gbk, .gbk.gz]&#34;, type=[&#34;.gbk&#34;, &#34;.gz&#34;,&#34;.gbff&#34;], accept_multiple_files=True)
    genome = list( map(lambda x: x.getvalue(), multiple_files_gbk))

    multiple_files_fasta = st.sidebar.file_uploader(&#34;Upload one or more fasta file [ .fasta, .fasta.gz]&#34;, type=[&#34;.fasta&#34;, &#34;.gz&#34;,&#34;.fna&#34;], accept_multiple_files=True)
    fasta = list( map(lambda x: x.getvalue(), multiple_files_fasta))


    multiple_files_gff= st.sidebar.file_uploader(&#34;Upload gff/gtf file if you are using fasta [ .gff, .gtf]&#34;, type=[&#34;.gff&#34;, &#34;.gtf&#34;], accept_multiple_files=True)
    gff = list( map(lambda x: x.getvalue(), multiple_files_gff))

    DemoGenome = st.sidebar.selectbox(&#34;OR Use Demo GBK&#34;,[&#39;Carsonella_ruddii.gbk.gz&#39;,&#39;Pseudomonas_aeruginosa.gbk.gz&#39;])
    DemoGenomepath = os.path.join(DATA_DIR,DemoGenome)
    demo = open(DemoGenomepath,&#34;rb&#34;)
    #st.write(&#34;You selected this option &#34;,xx)

  

    pam = st.sidebar.text_input(&#34;Input PAM Motif [ E.g. NGG ] &#34;, &#34;NGG&#34;)
    restriction_enzyme_list = st_tags_sidebar(label = &#39;Restriction Enzymes[e.g. NGRT]:&#39;,
                                                                  text  = &#39;Enter to add more&#39;, 
                                                                  value = [&#39;NGRT&#39;])
    #restriction_enzyme_list= st.sidebar.text_input(&#34;Restriction Enzymes list [ E.g. NGRT ] &#34;, &#34;NGRT&#34;)
    pam_orientation = st.sidebar.selectbox(
        &#34;PAM Orientation [ Options: 3prime, 5prime ]&#34;, (&#34;3prime&#34;, &#34;5prime&#34;))
    guidelength = st.sidebar.number_input(&#39;Guidelength [ Options: 10 - 27 ]&#39;, 10, 27, value=20)
    lsr = st.sidebar.number_input(&#39;Length of seed region[ Options: 0 - 27 ]&#39;, 0, 27, value=10)
    #dtype = st.sidebar.selectbox(
    #  &#34;Type of Edit Distance [ Options: hamming, leven ]&#34;, (&#34;hamming&#34;, &#34;leven&#34;))
    dist = st.sidebar.number_input(&#39;Edit Distance [Options: 0 - 5 ]&#39;, 0, 5, value=2)
    before = st.sidebar.number_input(&#39;Before [Options: 1 - 500 ]&#39;, 1, 500, value=100, step=50)
    into = st.sidebar.number_input(&#39;Into [Options: 1 - 500 ]&#39;, 1, 500, value=200, step=50)
    knum = st.sidebar.number_input(&#39;Similar Guides[Options: 2 - 20 ]&#39;, 2, 20, value=3)
    controls = st.sidebar.number_input(&#39;Control RNAs&#39;, 1, 1000, value=10, step=100)
    #threads = st.sidebar.number_input(&#39;Threads [ Options: 2, 4, 6, 8]&#39;, 2, 8, step=2)
                                                                  

    if demo and &#34;input.gbk&#34;:
        with genome_connect(demo):
            #st.write(&#34;Connection object:&#34;, conn)
            args = [&#34;guidemaker&#34;,
            &#34;-i&#34;, &#34;input.gbk&#34;,
            &#34;-p&#34;, pam,
            &#34;--guidelength&#34;, str(guidelength),
            &#34;--pam_orientation&#34;, pam_orientation,
            &#34;--lsr&#34;, str(lsr),
            &#34;--dtype&#34;, str(&#34;hamming&#34;),
            &#34;--dist&#34;, str(dist),
            &#34;--outdir&#34;, sessionID,
            &#34;--log&#34;, logfilename,
            &#34;--into&#34;, str(into),
            &#34;--before&#34;, str(before),
            &#34;--knum&#34;, str(knum),
            &#34;--controls&#34;, str(controls),
            &#34;--threads&#34;, str(2),
            &#34;--cfd_score&#34;,
            &#34;--doench_efficiency_score&#34;,
            &#34;--restriction_enzyme_list&#34;]
            scriptorun = args + restriction_enzyme_list

    if genome and &#34;input.gbk&#34;:
        with genome_connect(genome):
            #st.write(&#34;Connection object:&#34;, conn)
            args = [&#34;guidemaker&#34;,
            &#34;-i&#34;, &#34;input.gbk&#34;,
            &#34;-p&#34;, pam,
            &#34;--guidelength&#34;, str(guidelength),
            &#34;--pam_orientation&#34;, pam_orientation,
            &#34;--lsr&#34;, str(lsr),
             &#34;--dtype&#34;, str(&#34;hamming&#34;),
            &#34;--dist&#34;, str(dist),
            &#34;--outdir&#34;, sessionID,
            &#34;--log&#34;, logfilename,
            &#34;--into&#34;, str(into),
            &#34;--before&#34;, str(before),
            &#34;--knum&#34;, str(knum),
            &#34;--controls&#34;, str(controls),
            &#34;--threads&#34;, str(2),
            &#34;--cfd_score&#34;,
            &#34;--doench_efficiency_score&#34;,
            &#34;--restriction_enzyme_list&#34;]
            scriptorun = args + restriction_enzyme_list

    if fasta and gff:
        with fasta_connect(fasta):
            with gff_connect(gff):
                #st.write(&#34;Connection object:&#34;, conn)
                args = [&#34;guidemaker&#34;,
                &#34;-f&#34;, &#34;input.fasta&#34;,
                &#34;-g&#34;, &#34;input.gff&#34;,
                &#34;-p&#34;, pam,
                &#34;--guidelength&#34;, str(guidelength),
                &#34;--pam_orientation&#34;, pam_orientation,
                &#34;--lsr&#34;, str(lsr),
                &#34;--dtype&#34;, str(&#34;hamming&#34;),
                &#34;--dist&#34;, str(dist),
                &#34;--outdir&#34;, sessionID,
                &#34;--log&#34;, logfilename,
                &#34;--into&#34;, str(into),
                &#34;--before&#34;, str(before),
                &#34;--knum&#34;, str(knum),
                &#34;--controls&#34;, str(controls),
                &#34;--threads&#34;, str(2),
                &#34;--cfd_score&#34;,
                &#34;--doench_efficiency_score&#34;,
                &#34;--restriction_enzyme_list&#34;]
                scriptorun = args + restriction_enzyme_list

    if(st.sidebar.button(&#34;SUBMIT&#34;)):
        # st.markdown(&#34;&#34;&#34;🏃🏃🏃🏃🏃🏃🏃🏃&#34;&#34;&#34;)
        run_command(scriptorun)
    

    if os.path.exists(sessionID):

        #source = pd.read_csv(os.path.join(&#34;./&#34;, sessionID,&#39;targets.csv&#39;))
        source = pd.read_csv(os.path.join(&#34;./&#34;, sessionID, &#39;targets.csv&#39;), low_memory=False)

        accession_list = list(set(source[&#39;Accession&#39;]))
        for accession in accession_list:
            accession_df = source[source[&#34;Accession&#34;] == accession]
            accession_info = f&#34;**Accession:** {accession}&#34;
            st.markdown(accession_info)
            st.write(guidemakerplot(accession_df))


        # Targets
        target_tab = &#34;✅ [Target Data](downloads/targets.csv)&#34;
        targets = pd.read_csv(os.path.join(&#34;./&#34;, sessionID, &#39;targets.csv&#39;), low_memory=False)
        targets.to_csv(str(DOWNLOADS_PATH / &#34;targets.csv&#34;), index=False)

        # Controls
        control_tab = &#34;✅ [Control Data](downloads/controls.csv)&#34;
        controls = pd.read_csv(os.path.join(&#34;./&#34;, sessionID, &#39;controls.csv&#39;), low_memory=False)
        controls.to_csv(str(DOWNLOADS_PATH / &#34;controls.csv&#34;), index=False)

        # logs
        with st.expander(&#34;Results&#34;):
            st.write(target_tab)
            st.write(control_tab)
            st.write(get_binary_file_downloader_html(
                logfilename, &#39;✅ Log File&#39;), unsafe_allow_html=True)
    else:
        pass

    # Parameters Dictionary
    image = Image.open(guidemaker.APP_PARAMETER_IMG)
    optionals = st.expander(&#34;Parameter Dictionary&#34;, False)
    optionals.image(image, caption=&#39;GuideMaker Parameters&#39;, use_column_width=True)

    with st.expander(&#34;Designing Experiments with GuideMaker Results&#34;):
        intro_markdown = read_markdown_file(guidemaker.APP_EXPERIMENT_FILE)
        st.markdown(intro_markdown, unsafe_allow_html=True)

    st.markdown(&#34;&#34;&#34;
    ##### API documentation 📖

    API documentation for the module can be found [here](https://github.com/USDA-ARS-GBRU/GuideMaker)


    ##### License information ©️

    *Guidemaker was created by the United States Department of Agriculture - Agricultural Research Service (USDA-ARS). As a work of the United States Government this software is available under the CC0 1.0 Universal Public Domain Dedication (CC0 1.0)*

    &#34;&#34;&#34;)
    

    # Check if the output dir exist, if yes, delete so that previous results are not display
    try:
        shutil.rmtree(sessionID, ignore_errors=True)
        os.remove(logfilename)
        os.remove(&#39;input.gbk&#39;)
        os.remove(&#39;input.fasta&#39;)
        os.remove(&#39;input.gff&#39;)
    except FileNotFoundError as e:
        pass

if __name__ == &#34;__main__&#34;:
    main()</code></pre>
</details>
</section>
<section>
</section>
<section>
</section>
<section>
<h2 class="section-title" id="header-functions">Functions</h2>
<dl>
<dt id="guidemaker.data.app.fasta_connect"><code class="name flex">
<span>def <span class="ident">fasta_connect</span></span>(<span>db_bytes)</span>
</code></dt>
<dd>
<div class="desc"><p>Write input genome to local disk and clean after using.</p></div>
<details class="source">
<summary>
<span>Expand source code</span>
</summary>
<pre><code class="python">@contextmanager
def fasta_connect(db_bytes):
    &#34;&#34;&#34;Write input genome to local disk and clean after using.&#34;&#34;&#34;
    fp = Path(str(uuid4()))
    with open(&#39;input.fasta&#39;, &#39;wb&#39;) as file:
        for i in db_bytes:
            fp = bytearray(i)
            file.write(fp)
    conn = str(fp)
    try:
        yield conn
    finally:
        pass</code></pre>
</details>
</dd>
<dt id="guidemaker.data.app.genome_connect"><code class="name flex">
<span>def <span class="ident">genome_connect</span></span>(<span>db_bytes)</span>
</code></dt>
<dd>
<div class="desc"><p>Write input genome to local disk and clean after using.</p></div>
<details class="source">
<summary>
<span>Expand source code</span>
</summary>
<pre><code class="python">@contextmanager
def genome_connect(db_bytes):
    &#34;&#34;&#34;Write input genome to local disk and clean after using.&#34;&#34;&#34;
    fp = Path(str(uuid4()))
    with open(&#39;input.gbk&#39;, &#39;wb&#39;) as file:
        for i in db_bytes:
            fp = bytearray(i)
            file.write(fp)
    conn = str(fp)
    try:
        yield conn
    finally:
        pass</code></pre>
</details>
</dd>
<dt id="guidemaker.data.app.get_binary_file_downloader_html"><code class="name flex">
<span>def <span class="ident">get_binary_file_downloader_html</span></span>(<span>bin_file, file_label='File')</span>
</code></dt>
<dd>
<div class="desc"><p>Binary file downloader in html format.</p></div>
<details class="source">
<summary>
<span>Expand source code</span>
</summary>
<pre><code class="python">def get_binary_file_downloader_html(bin_file, file_label=&#39;File&#39;):
    &#34;&#34;&#34;Binary file downloader in html format.&#34;&#34;&#34;
    with open(bin_file, &#39;rb&#39;) as f:
        data = f.read()
    bin_str = base64.b64encode(data).decode()
    href = f&#39;&lt;a href=&#34;data:application/octet-stream;base64,{bin_str}&#34; download=&#34;{os.path.basename(bin_file)}&#34;&gt;{file_label}&lt;/a&gt;&#39;
    return href</code></pre>
</details>
</dd>
<dt id="guidemaker.data.app.gff_connect"><code class="name flex">
<span>def <span class="ident">gff_connect</span></span>(<span>db_bytes)</span>
</code></dt>
<dd>
<div class="desc"><p>Write input genome to local disk and clean after using.</p></div>
<details class="source">
<summary>
<span>Expand source code</span>
</summary>
<pre><code class="python">@contextmanager
def gff_connect(db_bytes):
    &#34;&#34;&#34;Write input genome to local disk and clean after using.&#34;&#34;&#34;
    fp = Path(str(uuid4()))
    with open(&#39;input.gff&#39;, &#39;wb&#39;) as file:
        for i in db_bytes:
            fp = bytearray(i)
            file.write(fp)
    conn = str(fp)
    try:
        yield conn
    finally:
        pass</code></pre>
</details>
</dd>
<dt id="guidemaker.data.app.guidemakerplot"><code class="name flex">
<span>def <span class="ident">guidemakerplot</span></span>(<span>df)</span>
</code></dt>
<dd>
<div class="desc"><p>Returns guidemaker plot describing PAM targets.</p></div>
<details class="source">
<summary>
<span>Expand source code</span>
</summary>
<pre><code class="python">def guidemakerplot(df):
    &#34;&#34;&#34;Returns guidemaker plot describing PAM targets.&#34;&#34;&#34;
    source = df
    brush = alt.selection(type=&#39;interval&#39;, encodings=[&#39;x&#39;])
    binnum = int(round(source[&#39;Feature end&#39;].max() / 200, 0))
    display_info = source.columns.tolist()

    # Feature density
    densityF = alt.Chart(source).transform_density(
    &#39;Feature start&#39;,
    as_=[&#39;Feature start&#39;, &#39;Feature Density&#39;],
    extent=[1, source[&#39;Feature end&#39;].max()],
    bandwidth=binnum,
    ).mark_area(color=&#39;black&#39;, opacity=0.6).encode(
    x=alt.X(&#39;Feature start&#39;, axis=alt.Axis(title=&#39;Genome Coordinates (bp)&#39;, tickCount=5)),
    y=&#39;Feature Density:Q&#39;,
    ).properties(height=50)

    # gRNA density
    densityg = alt.Chart(source).transform_density(
    &#39;Guide start&#39;,
    as_=[&#39;Guide start&#39;, &#39;Guide Density&#39;],
    extent=[1, source[&#39;Feature end&#39;].max()],
    bandwidth=binnum,
    ).mark_area(color=&#39;pink&#39;, opacity=0.6).encode(
    x=alt.X(&#39;Guide start&#39;, axis=alt.Axis(title=&#39;Genome Coordinates (bp)&#39;, tickCount=5)),
    y=&#39;Guide Density:Q&#39;,
    ).properties(height=50).add_selection(brush)

    # locus tag
    locus = alt.Chart(source).mark_bar(cornerRadiusTopLeft=3, cornerRadiusTopRight=3).encode(
    x=&#39;count(locus_tag):Q&#39;,
    y=alt.Y(&#39;locus_tag&#39;, axis=alt.Axis(title=&#39;Locus&#39;)),
    color=&#39;PAM:N&#39;,
    tooltip=display_info
    ).transform_filter(
    brush
    ).interactive().properties(height=500)

    gmplot = densityF &amp; densityg &amp; locus
    return gmplot</code></pre>
</details>
</dd>
<dt id="guidemaker.data.app.main"><code class="name flex">
<span>def <span class="ident">main</span></span>(<span>arglist: list = None)</span>
</code></dt>
<dd>
<div class="desc"><p>Run web App.</p></div>
<details class="source">
<summary>
<span>Expand source code</span>
</summary>
<pre><code class="python">def main(arglist: list = None):
    &#34;&#34;&#34;Run web App.&#34;&#34;&#34;
    header = &#34;GuideMaker&#34;
    subheader = &#34;Software to design CRISPR-Cas guide RNA pools in non-model genomes 🦠 🧬&#34;
    st.markdown(f&#39;&lt;strong style=&#34;font-family:Hoefler Text;font-size: 36px;color: #0021A5&#34;&gt;{header}&lt;/strong&gt;&#39;,
                unsafe_allow_html=True)
    st.markdown(
        f&#39;&lt;strong style=&#34;font-family:Hoefler Text;font-size: 18px;color: #FA4616&#34;&gt;{subheader}&lt;/strong&gt;&#39;, unsafe_allow_html=True)

    st.markdown(&#34;---&#34;)

    sessionID = str(uuid.uuid1())
    # st.write(sessionID)
    logfilename = sessionID + &#34;_log.txt&#34;

    # Create a downloads directory within the streamlit static asset directory
    # and write output files to it. Binary downloader as html file  has limited
    # file size for conversion. So, we need to write it to local folder.
    STREAMLIT_STATIC_PATH = pathlib.Path(st.__path__[0]) / &#39;static&#39;
    DOWNLOADS_PATH = (STREAMLIT_STATIC_PATH / &#34;downloads&#34;)
    if not DOWNLOADS_PATH.is_dir():
        DOWNLOADS_PATH.mkdir()

    # Define input parameters and widgets

    multiple_files_gbk = st.sidebar.file_uploader(&#34;Upload one or more Genome file [ .gbk, .gbk.gz]&#34;, type=[&#34;.gbk&#34;, &#34;.gz&#34;,&#34;.gbff&#34;], accept_multiple_files=True)
    genome = list( map(lambda x: x.getvalue(), multiple_files_gbk))

    multiple_files_fasta = st.sidebar.file_uploader(&#34;Upload one or more fasta file [ .fasta, .fasta.gz]&#34;, type=[&#34;.fasta&#34;, &#34;.gz&#34;,&#34;.fna&#34;], accept_multiple_files=True)
    fasta = list( map(lambda x: x.getvalue(), multiple_files_fasta))


    multiple_files_gff= st.sidebar.file_uploader(&#34;Upload gff/gtf file if you are using fasta [ .gff, .gtf]&#34;, type=[&#34;.gff&#34;, &#34;.gtf&#34;], accept_multiple_files=True)
    gff = list( map(lambda x: x.getvalue(), multiple_files_gff))

    DemoGenome = st.sidebar.selectbox(&#34;OR Use Demo GBK&#34;,[&#39;Carsonella_ruddii.gbk.gz&#39;,&#39;Pseudomonas_aeruginosa.gbk.gz&#39;])
    DemoGenomepath = os.path.join(DATA_DIR,DemoGenome)
    demo = open(DemoGenomepath,&#34;rb&#34;)
    #st.write(&#34;You selected this option &#34;,xx)

  

    pam = st.sidebar.text_input(&#34;Input PAM Motif [ E.g. NGG ] &#34;, &#34;NGG&#34;)
    restriction_enzyme_list = st_tags_sidebar(label = &#39;Restriction Enzymes[e.g. NGRT]:&#39;,
                                                                  text  = &#39;Enter to add more&#39;, 
                                                                  value = [&#39;NGRT&#39;])
    #restriction_enzyme_list= st.sidebar.text_input(&#34;Restriction Enzymes list [ E.g. NGRT ] &#34;, &#34;NGRT&#34;)
    pam_orientation = st.sidebar.selectbox(
        &#34;PAM Orientation [ Options: 3prime, 5prime ]&#34;, (&#34;3prime&#34;, &#34;5prime&#34;))
    guidelength = st.sidebar.number_input(&#39;Guidelength [ Options: 10 - 27 ]&#39;, 10, 27, value=20)
    lsr = st.sidebar.number_input(&#39;Length of seed region[ Options: 0 - 27 ]&#39;, 0, 27, value=10)
    #dtype = st.sidebar.selectbox(
    #  &#34;Type of Edit Distance [ Options: hamming, leven ]&#34;, (&#34;hamming&#34;, &#34;leven&#34;))
    dist = st.sidebar.number_input(&#39;Edit Distance [Options: 0 - 5 ]&#39;, 0, 5, value=2)
    before = st.sidebar.number_input(&#39;Before [Options: 1 - 500 ]&#39;, 1, 500, value=100, step=50)
    into = st.sidebar.number_input(&#39;Into [Options: 1 - 500 ]&#39;, 1, 500, value=200, step=50)
    knum = st.sidebar.number_input(&#39;Similar Guides[Options: 2 - 20 ]&#39;, 2, 20, value=3)
    controls = st.sidebar.number_input(&#39;Control RNAs&#39;, 1, 1000, value=10, step=100)
    #threads = st.sidebar.number_input(&#39;Threads [ Options: 2, 4, 6, 8]&#39;, 2, 8, step=2)
                                                                  

    if demo and &#34;input.gbk&#34;:
        with genome_connect(demo):
            #st.write(&#34;Connection object:&#34;, conn)
            args = [&#34;guidemaker&#34;,
            &#34;-i&#34;, &#34;input.gbk&#34;,
            &#34;-p&#34;, pam,
            &#34;--guidelength&#34;, str(guidelength),
            &#34;--pam_orientation&#34;, pam_orientation,
            &#34;--lsr&#34;, str(lsr),
            &#34;--dtype&#34;, str(&#34;hamming&#34;),
            &#34;--dist&#34;, str(dist),
            &#34;--outdir&#34;, sessionID,
            &#34;--log&#34;, logfilename,
            &#34;--into&#34;, str(into),
            &#34;--before&#34;, str(before),
            &#34;--knum&#34;, str(knum),
            &#34;--controls&#34;, str(controls),
            &#34;--threads&#34;, str(2),
            &#34;--cfd_score&#34;,
            &#34;--doench_efficiency_score&#34;,
            &#34;--restriction_enzyme_list&#34;]
            scriptorun = args + restriction_enzyme_list

    if genome and &#34;input.gbk&#34;:
        with genome_connect(genome):
            #st.write(&#34;Connection object:&#34;, conn)
            args = [&#34;guidemaker&#34;,
            &#34;-i&#34;, &#34;input.gbk&#34;,
            &#34;-p&#34;, pam,
            &#34;--guidelength&#34;, str(guidelength),
            &#34;--pam_orientation&#34;, pam_orientation,
            &#34;--lsr&#34;, str(lsr),
             &#34;--dtype&#34;, str(&#34;hamming&#34;),
            &#34;--dist&#34;, str(dist),
            &#34;--outdir&#34;, sessionID,
            &#34;--log&#34;, logfilename,
            &#34;--into&#34;, str(into),
            &#34;--before&#34;, str(before),
            &#34;--knum&#34;, str(knum),
            &#34;--controls&#34;, str(controls),
            &#34;--threads&#34;, str(2),
            &#34;--cfd_score&#34;,
            &#34;--doench_efficiency_score&#34;,
            &#34;--restriction_enzyme_list&#34;]
            scriptorun = args + restriction_enzyme_list

    if fasta and gff:
        with fasta_connect(fasta):
            with gff_connect(gff):
                #st.write(&#34;Connection object:&#34;, conn)
                args = [&#34;guidemaker&#34;,
                &#34;-f&#34;, &#34;input.fasta&#34;,
                &#34;-g&#34;, &#34;input.gff&#34;,
                &#34;-p&#34;, pam,
                &#34;--guidelength&#34;, str(guidelength),
                &#34;--pam_orientation&#34;, pam_orientation,
                &#34;--lsr&#34;, str(lsr),
                &#34;--dtype&#34;, str(&#34;hamming&#34;),
                &#34;--dist&#34;, str(dist),
                &#34;--outdir&#34;, sessionID,
                &#34;--log&#34;, logfilename,
                &#34;--into&#34;, str(into),
                &#34;--before&#34;, str(before),
                &#34;--knum&#34;, str(knum),
                &#34;--controls&#34;, str(controls),
                &#34;--threads&#34;, str(2),
                &#34;--cfd_score&#34;,
                &#34;--doench_efficiency_score&#34;,
                &#34;--restriction_enzyme_list&#34;]
                scriptorun = args + restriction_enzyme_list

    if(st.sidebar.button(&#34;SUBMIT&#34;)):
        # st.markdown(&#34;&#34;&#34;🏃🏃🏃🏃🏃🏃🏃🏃&#34;&#34;&#34;)
        run_command(scriptorun)
    

    if os.path.exists(sessionID):

        #source = pd.read_csv(os.path.join(&#34;./&#34;, sessionID,&#39;targets.csv&#39;))
        source = pd.read_csv(os.path.join(&#34;./&#34;, sessionID, &#39;targets.csv&#39;), low_memory=False)

        accession_list = list(set(source[&#39;Accession&#39;]))
        for accession in accession_list:
            accession_df = source[source[&#34;Accession&#34;] == accession]
            accession_info = f&#34;**Accession:** {accession}&#34;
            st.markdown(accession_info)
            st.write(guidemakerplot(accession_df))


        # Targets
        target_tab = &#34;✅ [Target Data](downloads/targets.csv)&#34;
        targets = pd.read_csv(os.path.join(&#34;./&#34;, sessionID, &#39;targets.csv&#39;), low_memory=False)
        targets.to_csv(str(DOWNLOADS_PATH / &#34;targets.csv&#34;), index=False)

        # Controls
        control_tab = &#34;✅ [Control Data](downloads/controls.csv)&#34;
        controls = pd.read_csv(os.path.join(&#34;./&#34;, sessionID, &#39;controls.csv&#39;), low_memory=False)
        controls.to_csv(str(DOWNLOADS_PATH / &#34;controls.csv&#34;), index=False)

        # logs
        with st.expander(&#34;Results&#34;):
            st.write(target_tab)
            st.write(control_tab)
            st.write(get_binary_file_downloader_html(
                logfilename, &#39;✅ Log File&#39;), unsafe_allow_html=True)
    else:
        pass

    # Parameters Dictionary
    image = Image.open(guidemaker.APP_PARAMETER_IMG)
    optionals = st.expander(&#34;Parameter Dictionary&#34;, False)
    optionals.image(image, caption=&#39;GuideMaker Parameters&#39;, use_column_width=True)

    with st.expander(&#34;Designing Experiments with GuideMaker Results&#34;):
        intro_markdown = read_markdown_file(guidemaker.APP_EXPERIMENT_FILE)
        st.markdown(intro_markdown, unsafe_allow_html=True)

    st.markdown(&#34;&#34;&#34;
    ##### API documentation 📖

    API documentation for the module can be found [here](https://github.com/USDA-ARS-GBRU/GuideMaker)


    ##### License information ©️

    *Guidemaker was created by the United States Department of Agriculture - Agricultural Research Service (USDA-ARS). As a work of the United States Government this software is available under the CC0 1.0 Universal Public Domain Dedication (CC0 1.0)*

    &#34;&#34;&#34;)
    

    # Check if the output dir exist, if yes, delete so that previous results are not display
    try:
        shutil.rmtree(sessionID, ignore_errors=True)
        os.remove(logfilename)
        os.remove(&#39;input.gbk&#39;)
        os.remove(&#39;input.fasta&#39;)
        os.remove(&#39;input.gff&#39;)
    except FileNotFoundError as e:
        pass</code></pre>
</details>
</dd>
<dt id="guidemaker.data.app.read_markdown_file"><code class="name flex">
<span>def <span class="ident">read_markdown_file</span></span>(<span>markdown_file)</span>
</code></dt>
<dd>
<div class="desc"><p>Read markdown file.</p></div>
<details class="source">
<summary>
<span>Expand source code</span>
</summary>
<pre><code class="python">def read_markdown_file(markdown_file):
    &#34;&#34;&#34;Read markdown file.&#34;&#34;&#34;
    return Path(markdown_file).read_text()</code></pre>
</details>
</dd>
<dt id="guidemaker.data.app.run_command"><code class="name flex">
<span>def <span class="ident">run_command</span></span>(<span>args)</span>
</code></dt>
<dd>
<div class="desc"><p>Run command, transfer stdout/stderr back into Streamlit and manage error.</p></div>
<details class="source">
<summary>
<span>Expand source code</span>
</summary>
<pre><code class="python">@st.cache(suppress_st_warning=True)
def run_command(args):
    &#34;&#34;&#34;Run command, transfer stdout/stderr back into Streamlit and manage error.&#34;&#34;&#34;
    st.info(f&#34;Running:: &#39;{&#39; &#39;.join(args)}&#39;&#34;)
    result = subprocess.run(args, capture_output=True, text=True)
    try:
        result.check_returncode()
        # st.info(result.stdout)
        # st.text(result.stderr)
    except subprocess.CalledProcessError as e:
        st.error(result.stderr)
        raise e</code></pre>
</details>
</dd>
</dl>
</section>
<section>
</section>
</article>
<nav id="sidebar">
<h1>Index</h1>
<div class="toc">
<ul></ul>
</div>
<ul id="index">
<li><h3>Super-module</h3>
<ul>
<li><code><a title="guidemaker.data" href="index.html">guidemaker.data</a></code></li>
</ul>
</li>
<li><h3><a href="#header-functions">Functions</a></h3>
<ul class="">
<li><code><a title="guidemaker.data.app.fasta_connect" href="#guidemaker.data.app.fasta_connect">fasta_connect</a></code></li>
<li><code><a title="guidemaker.data.app.genome_connect" href="#guidemaker.data.app.genome_connect">genome_connect</a></code></li>
<li><code><a title="guidemaker.data.app.get_binary_file_downloader_html" href="#guidemaker.data.app.get_binary_file_downloader_html">get_binary_file_downloader_html</a></code></li>
<li><code><a title="guidemaker.data.app.gff_connect" href="#guidemaker.data.app.gff_connect">gff_connect</a></code></li>
<li><code><a title="guidemaker.data.app.guidemakerplot" href="#guidemaker.data.app.guidemakerplot">guidemakerplot</a></code></li>
<li><code><a title="guidemaker.data.app.main" href="#guidemaker.data.app.main">main</a></code></li>
<li><code><a title="guidemaker.data.app.read_markdown_file" href="#guidemaker.data.app.read_markdown_file">read_markdown_file</a></code></li>
<li><code><a title="guidemaker.data.app.run_command" href="#guidemaker.data.app.run_command">run_command</a></code></li>
</ul>
</li>
</ul>
</nav>
</main>
<footer id="footer">
<p>Generated by <a href="https://pdoc3.github.io/pdoc" title="pdoc: Python API documentation generator"><cite>pdoc</cite> 0.10.0</a>.</p>
</footer>
</body>
</html>